File: NoriTests/ClipClassifierTests.swift
import AppKit
import Foundation
import Testing
@testable import Nori

@Suite("ClipClassifier")
struct ClipClassifierTests {
    private func snapshot(
        _ items: [[(String, Data?)]],
        declared: Set<String>? = nil,
        app: String? = "com.apple.TextEdit",
        appName: String? = "TextEdit",
        changeCount: Int = 1
    ) -> PasteboardSnapshot {
        let built = items.map { PasteboardSnapshot.Item(representations: $0.map { ($0.0, $0.1) }) }
        let declaredTypes = declared ?? Set(built.flatMap(\.types))
        return PasteboardSnapshot(changeCount: changeCount, declaredTypes: declaredTypes, items: built,
                                  sourceBundleID: app, sourceAppName: appName)
    }

    private func text(_ string: String) -> (String, Data?) {
        (PasteboardType.utf8PlainText, Data(string.utf8))
    }

    private func draft(_ outcome: CaptureOutcome) throws -> ClipDraft {
        guard case let .captured(draft) = outcome else {
            Issue.record("expected capture, got \(outcome)")
            throw CancellationError()
        }
        return draft
    }

    @Test func plainText() throws {
        let draft = try draft(ClipClassifier.classify(snapshot([[text("  Hello, Nori!\n")]])))
        #expect(draft.kind == .text)
        #expect(draft.title == "Hello, Nori!")
        #expect(draft.searchText == "  Hello, Nori!\n")
        #expect(draft.characterCount == 15)
        #expect(draft.lineCount == 1)
        #expect(draft.sourceBundleID == "com.apple.TextEdit")
        #expect(draft.sourceAppName == "TextEdit")
        #expect(draft.contents.count == 1)
        #expect(!draft.isRichText)
    }

    @Test func link() throws {
        let draft = try draft(ClipClassifier.classify(snapshot([[text("https://maccy.app/\n")]])))
        #expect(draft.kind == .link)
        #expect(draft.linkURL?.host() == "maccy.app")
    }

    @Test func color() throws {
        let draft = try draft(ClipClassifier.classify(snapshot([[text("#ff8800")]])))
        #expect(draft.kind == .color)
        #expect(draft.colorHex == "#FF8800")
    }

    @Test func code() throws {
        let source = "func hello() {\n    print(\"hi\")\n}\n"
        let draft = try draft(ClipClassifier.classify(snapshot([[text(source)]])))
        #expect(draft.kind == .code)
    }

    @Test func codeFromEditorNeedsLessEvidence() throws {
        let source = "let a = 1\nlet b = 2\n"
        #expect(try draft(ClipClassifier.classify(snapshot([[text(source)]], app: "com.apple.dt.Xcode"))).kind == .code)
        #expect(try draft(ClipClassifier.classify(snapshot([[text(source)]], app: "com.apple.Notes"))).kind == .text)
    }

    @Test func richTextIsAFlagOnText() throws {
        let attributed = NSMutableAttributedString(string: "Bold and plain")
        attributed.addAttribute(.font, value: NSFont.boldSystemFont(ofSize: 12), range: NSRange(location: 0, length: 4))
        let rtf = try #require(attributed.rtf(from: NSRange(location: 0, length: attributed.length), documentAttributes: [:]))
        let draft = try draft(ClipClassifier.classify(snapshot([[text("Bold and plain"), (PasteboardType.rtf, rtf)]])))
        #expect(draft.kind == .text)
        #expect(draft.isRichText)
        #expect(draft.title == "Bold and plain")
        #expect(draft.contents.map(\.type) == [PasteboardType.utf8PlainText, PasteboardType.rtf])
    }

    @Test func uniformRTFIsNotRich() throws {
        let attributed = NSAttributedString(string: "just text", attributes: [.font: NSFont.systemFont(ofSize: 12)])
        let rtf = try #require(attributed.rtf(from: NSRange(location: 0, length: attributed.length), documentAttributes: [:]))
        let draft = try draft(ClipClassifier.classify(snapshot([[text("just text"), (PasteboardType.rtf, rtf)]])))
        #expect(!draft.isRichText)
    }

    @Test func imageKeepsPNGOnlyAndThumbnail() throws {
        let png = try #require(TestImages.png(width: 300, height: 120))
        let tiff = try #require(NSImage(data: png)?.tiffRepresentation)
        let draft = try draft(ClipClassifier.classify(snapshot([[(PasteboardType.tiff, tiff), (PasteboardType.png, png), text("alt text")]])))
        #expect(draft.kind == .image)
        #expect(draft.imagePixelSize == CGSize(width: 300, height: 120))
        #expect(draft.title == "Image 300×120")
        #expect(draft.searchText == "alt text")
        #expect(draft.contents.map(\.type) == [PasteboardType.png, PasteboardType.utf8PlainText])
        let thumbnail = try #require(draft.thumbnail)
        let size = try #require(ImageNormalizer.pixelSize(of: thumbnail))
        #expect(size.width <= 224 && size.height <= 224)
    }

    @Test func tiffOnlyIsTranscodedToPNG() throws {
        let png = try #require(TestImages.png(width: 40, height: 30))
        let tiff = try #require(NSImage(data: png)?.tiffRepresentation)
        let draft = try draft(ClipClassifier.classify(snapshot([[(PasteboardType.tiff, tiff)]])))
        #expect(draft.contents.map(\.type) == [PasteboardType.png])
        #expect(draft.imagePixelSize == CGSize(width: 40, height: 30))
    }

    @Test func files() throws {
        let url = URL(fileURLWithPath: "/tmp/My File.txt")
        let other = URL(fileURLWithPath: "/tmp/other.pdf")
        let snap = snapshot([
            [(PasteboardType.fileURL, url.dataRepresentation), text("My File.txt")],
            [(PasteboardType.fileURL, other.dataRepresentation), text("other.pdf")],
        ])
        let draft = try draft(ClipClassifier.classify(snap))
        #expect(draft.kind == .file)
        #expect(draft.fileURLs.map(\.lastPathComponent) == ["My File.txt", "other.pdf"])
        #expect(draft.title == "My File.txt\nother.pdf")
        #expect(draft.lineCount == 2)
    }

    @Test func mergesMultipleItemsWithoutDuplicatingTypes() throws {
        let snap = snapshot([[text("one")], [text("two"), (PasteboardType.html, Data("<b>two</b>".utf8))]])
        let draft = try draft(ClipClassifier.classify(snap))
        #expect(draft.contents.map(\.type) == [PasteboardType.utf8PlainText, PasteboardType.html])
        #expect(draft.title == "one")
    }

    @Test func whitespaceOnlyIsRejected() {
        #expect(ClipClassifier.classify(snapshot([[text("   \n\t")]])) == .rejected(.nothingToStore))
    }

    @Test func missingDataIsRejected() {
        #expect(ClipClassifier.classify(snapshot([[(PasteboardType.utf8PlainText, nil)]])) == .rejected(.nothingToStore))
    }

    @Test func concealedLeavesAGhostEvenWhenOnlyDeclared() {
        let snap = snapshot([[text("hunter2")]], declared: [PasteboardType.utf8PlainText, PasteboardType.concealed], appName: "1Password")
        #expect(ClipClassifier.classify(snap) == .ghost(.concealed(appName: "1Password")))
    }

    @Test func transientAndAutoGeneratedAreIgnored() {
        for type in [PasteboardType.transient, PasteboardType.autoGenerated] {
            let snap = snapshot([[text("x"), (type, Data())]])
            #expect(ClipClassifier.classify(snap) == .rejected(.ephemeral(type)))
        }
    }

    @Test func ignoredCustomTypes() {
        let snap = snapshot([[text("secret"), ("com.agilebits.onepassword", Data())]])
        #expect(ClipClassifier.classify(snap) == .rejected(.ignoredType("com.agilebits.onepassword")))
    }

    @Test func ignoredAppsIncludePasswordManagersByDefault() {
        let snap = snapshot([[text("pw")]], app: "com.apple.keychainaccess")
        #expect(ClipClassifier.classify(snap) == .rejected(.ignoredApp("com.apple.keychainaccess")))
    }

    @Test func ignoreRegexp() {
        var policy = CapturePolicy()
        policy.ignoreRegexps = ["^order-[0-9]{6}$"]
        #expect(ClipClassifier.classify(snapshot([[text("order-123456")]]), policy: policy) == .rejected(.matchedIgnoreRegexp))
        #expect(ClipClassifier.classify(snapshot([[text("order-12")]]), policy: policy) != .rejected(.matchedIgnoreRegexp))
    }

    @Test func fromNoriIsRejectedWithID() {
        let id = UUID()
        let snap = snapshot([[text("again"), (PasteboardType.noriItem, Data(id.uuidString.utf8))]])
        #expect(ClipClassifier.classify(snap) == .rejected(.fromNori(id)))
    }

    @Test func dropsDynamicAndMetadataTypes() throws {
        let snap = snapshot([[
            text("hello"),
            ("dyn.ah62d4rv4gu8y", Data([1])),
            ("com.microsoft.ole.source.abc", Data([2])),
            ("org.chromium.web-custom-data", Data([3])),
            ("com.apple.linkpresentation.metadata", Data([4])),
        ]])
        let draft = try draft(ClipClassifier.classify(snap))
        #expect(draft.contents.map(\.type) == [PasteboardType.utf8PlainText])
    }

    @Test func stripsWordLinkTypes() throws {
        let snap = snapshot([[
            text("Heading 1"),
            ("com.microsoft.ObjectLink", Data([1])),
            ("com.microsoft.Link-Source", Data([2])),
            (PasteboardType.pdf, Data([3])),
            (PasteboardType.rtf, Data("{\\rtf1 Heading 1}".utf8)),
        ]])
        let draft = try draft(ClipClassifier.classify(snap))
        #expect(!draft.contents.contains { $0.type == "com.microsoft.ObjectLink" || $0.type == PasteboardType.pdf })
        #expect(draft.contents.contains { $0.type == PasteboardType.rtf })
    }

    @Test func policyCanDisableImagesFilesAndText() throws {
        var policy = CapturePolicy()
        policy.captureImages = false
        policy.captureFiles = false
        let png = try #require(TestImages.png(width: 2, height: 2))
        let snap = snapshot([[(PasteboardType.png, png), (PasteboardType.rtf, Data([1])), text("caption")]])
        let draft = try draft(ClipClassifier.classify(snap, policy: policy))
        #expect(draft.kind == .text)
        #expect(draft.contents.map(\.type) == [PasteboardType.rtf, PasteboardType.utf8PlainText])

        policy.captureText = false
        #expect(ClipClassifier.classify(snapshot([[text("caption")]]), policy: policy) == .rejected(.nothingToStore))
    }

    @Test func oversizedImageLeavesAGhost() throws {
        var policy = CapturePolicy()
        policy.maxImageBytes = 4
        let png = try #require(TestImages.png(width: 2, height: 2))
        let outcome = ClipClassifier.classify(snapshot([[(PasteboardType.png, png), text("alt")]]), policy: policy)
        #expect(outcome == .ghost(.imageTooLarge(bytes: png.count)))
    }

    @Test func oversizedOtherRepresentationIsDroppedButClipKept() throws {
        var policy = CapturePolicy()
        policy.maxOtherRepresentationBytes = 4
        let snap = snapshot([[text("tiny"), ("com.apple.WebKit.custom-pasteboard-data", Data(repeating: 0, count: 10)), ("com.example.big", Data(repeating: 0, count: 10))]])
        let draft = try draft(ClipClassifier.classify(snap, policy: policy))
        #expect(draft.contents.map(\.type) == [PasteboardType.utf8PlainText])
    }

    @Test func hugeTextIsTruncated() throws {
        var policy = CapturePolicy()
        policy.maxTextBytes = 16
        let snap = snapshot([[text(String(repeating: "a", count: 100)), (PasteboardType.rtf, Data([1]))]])
        let draft = try draft(ClipClassifier.classify(snap, policy: policy))
        #expect(draft.isTruncated)
        #expect(draft.characterCount == 16)
        #expect(draft.contents.map(\.type) == [PasteboardType.utf8PlainText])
    }

    @Test func secretsGoToTheVault() throws {
        let outcome = ClipClassifier.classify(snapshot([[text("AKIAIOSFODNN7EXAMPLE")]]))
        guard case let .sensitive(sensitive) = outcome else {
            Issue.record("expected sensitive, got \(outcome)")
            return
        }
        #expect(sensitive.match == .awsAccessKey)
        #expect(sensitive.mask.hasSuffix("MPLE"))
        #expect(!sensitive.mask.contains("AKIA"))

        var policy = CapturePolicy()
        policy.maskSensitive = false
        #expect(try draft(ClipClassifier.classify(snapshot([[text("AKIAIOSFODNN7EXAMPLE")]]), policy: policy)).kind == .text)
    }

    @Test func hashIgnoresCustomTypesButNotContent() {
        let a = [ClipDraft.Content(type: PasteboardType.utf8PlainText, data: Data("x".utf8)),
                 ClipDraft.Content(type: "com.example.custom", data: Data([9]))]
        let b = [ClipDraft.Content(type: PasteboardType.utf8PlainText, data: Data("x".utf8))]
        let c = [ClipDraft.Content(type: PasteboardType.utf8PlainText, data: Data("y".utf8))]
        #expect(ClipClassifier.contentHash(of: a) == ClipClassifier.contentHash(of: b))
        #expect(ClipClassifier.contentHash(of: a) != ClipClassifier.contentHash(of: c))
    }

    @Test func universalClipboardIsFlaggedAndCanBeDisabled() throws {
        let snap = snapshot([[text("from iPhone")]], declared: [PasteboardType.utf8PlainText, PasteboardType.universalClipboard])
        let draft = try draft(ClipClassifier.classify(snap))
        #expect(draft.isFromUniversalClipboard)
        #expect(draft.sourceAppName == "iPhone or iPad")
        var policy = CapturePolicy()
        policy.captureUniversalClipboard = false
        #expect(ClipClassifier.classify(snap, policy: policy) == .rejected(.universalClipboardDisabled))
    }

    @Test func longTextIsTruncatedForTitleAndSearch() throws {
        let long = String(repeating: "a", count: 20_000)
        let draft = try draft(ClipClassifier.classify(snapshot([[text(long)]])))
        #expect(draft.title.count == ClipDraft.maxTitleLength)
        #expect(draft.searchText.count == ClipDraft.maxSearchTextLength)
        #expect(draft.characterCount == 20_000)
    }
}

File: windows/Nori.Core.Tests/ClipClassifierTests.cs
using System.Text;
using Nori.Core;
using Xunit;

namespace Nori.Core.Tests;

public class ClipClassifierTests
{
    private static ClipboardSnapshot Snapshot(
        IReadOnlyList<IReadOnlyList<(string Format, byte[]? Data)>> items,
        IReadOnlySet<string>? declared = null,
        string? app = "notepad.exe",
        string? appName = "Notepad",
        bool canIncludeInHistory = true)
    {
        var built = items.Select(i => new ClipboardSnapshot.Item(i)).ToList();
        var declaredFormats = declared ?? new HashSet<string>(built.SelectMany(i => i.Formats), StringComparer.OrdinalIgnoreCase);
        return new ClipboardSnapshot(declaredFormats, built, app, appName, canIncludeInHistory: canIncludeInHistory);
    }

    private static (string, byte[]?) Text(string s) => (ClipboardFormats.UnicodeText, Encoding.UTF8.GetBytes(s));

    private static ClipDraft Draft(CaptureOutcome outcome)
    {
        var captured = Assert.IsType<CaptureOutcome.Captured>(outcome);
        return captured.Draft;
    }

    [Fact]
    public void PlainText()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("  Hello, Nori!\n")]])));
        Assert.Equal(ClipKind.Text, draft.Kind);
        Assert.Equal("Hello, Nori!", draft.Title);
        Assert.Equal("  Hello, Nori!\n", draft.SearchText);
        Assert.Equal(15, draft.CharacterCount);
        Assert.Equal(1, draft.LineCount);
        Assert.Equal("notepad.exe", draft.SourceApp);
        Assert.Equal("Notepad", draft.SourceAppName);
        Assert.Single(draft.Contents);
        Assert.False(draft.IsRichText);
    }

    [Fact]
    public void Link()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("https://maccy.app/\n")]])));
        Assert.Equal(ClipKind.Link, draft.Kind);
        Assert.Equal("maccy.app", draft.LinkUrl?.Host);
    }

    [Fact]
    public void Color()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("#ff8800")]])));
        Assert.Equal(ClipKind.Color, draft.Kind);
        Assert.Equal("#FF8800", draft.ColorHex);
    }

    [Fact]
    public void Code()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("func hello() {\n    print(\"hi\")\n}\n")]])));
        Assert.Equal(ClipKind.Code, draft.Kind);
    }

    [Fact]
    public void CodeFromEditorNeedsLessEvidence()
    {
        const string source = "let a = 1\nlet b = 2\n";
        Assert.Equal(ClipKind.Code, Draft(ClipClassifier.Classify(Snapshot([[Text(source)]], app: "Code.exe"))).Kind);
        Assert.Equal(ClipKind.Text, Draft(ClipClassifier.Classify(Snapshot([[Text(source)]], app: "notepad.exe"))).Kind);
    }

    [Fact]
    public void RichTextIsAFlagOnText()
    {
        var rtf = Encoding.ASCII.GetBytes(@"{\rtf1\ansi{\fonttbl\f0\fswiss Helvetica;}\f0\fs24 \b Bold\b0  and plain}");
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("Bold and plain"), (ClipboardFormats.Rtf, rtf)]])));
        Assert.Equal(ClipKind.Text, draft.Kind);
        Assert.True(draft.IsRichText);
        Assert.Equal("Bold and plain", draft.Title);
        Assert.Equal([ClipboardFormats.UnicodeText, ClipboardFormats.Rtf], draft.Contents.Select(c => c.Format));
    }

    [Fact]
    public void UniformRtfIsNotRich()
    {
        var rtf = Encoding.ASCII.GetBytes(@"{\rtf1\ansi{\fonttbl\f0\fswiss Helvetica;}\f0\fs24 just text}");
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("just text"), (ClipboardFormats.Rtf, rtf)]])));
        Assert.False(draft.IsRichText);
    }

    [Fact]
    public void HtmlAloneNeverMakesAClipRich()
    {
        var html = HtmlFormat.Encode("<b>bold</b> text");
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text("bold text"), (ClipboardFormats.Html, html)]])));
        Assert.False(draft.IsRichText);
        Assert.Equal(2, draft.Contents.Count);
    }

    [Fact]
    public void ImageKeepsPngOnlyAndReadsSize()
    {
        var png = PngHeaderNormalizer.SolidPng(300, 120);
        var dib = new byte[64];
        var draft = Draft(ClipClassifier.Classify(Snapshot([[(ClipboardFormats.Dib, dib), (ClipboardFormats.Png, png), Text("alt text")]])));
        Assert.Equal(ClipKind.Image, draft.Kind);
        Assert.Equal((300, 120), draft.ImagePixelSize);
        Assert.Equal("Image 300×120", draft.Title);
        Assert.Equal("alt text", draft.SearchText);
        Assert.Equal([ClipboardFormats.Png, ClipboardFormats.UnicodeText], draft.Contents.Select(c => c.Format));
    }

    private sealed class StubNormalizer : IImageNormalizer
    {
        public ImageResult? Normalize(IReadOnlyList<ClipContent> contents) =>
            new(PngHeaderNormalizer.SolidPng(40, 30), 40, 30, PngHeaderNormalizer.SolidPng(20, 15));
    }

    [Fact]
    public void DibOnlyIsTranscodedByTheNormalizer()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[(ClipboardFormats.Dib, new byte[64])]]), normalizer: new StubNormalizer()));
        Assert.Equal([ClipboardFormats.Png], draft.Contents.Select(c => c.Format));
        Assert.Equal((40, 30), draft.ImagePixelSize);
        Assert.NotNull(draft.Thumbnail);
        Assert.Equal((20, 15), PngHeaderNormalizer.PngSize(draft.Thumbnail));
    }

    [Fact]
    public void Files()
    {
        var snap = Snapshot([
            [ClipDraft.File(@"C:\Users\me\Documents\My File.txt").ToTuple(), Text("My File.txt")],
            [ClipDraft.File(@"C:\tmp\other.pdf").ToTuple(), Text("other.pdf")],
        ]);
        var draft = Draft(ClipClassifier.Classify(snap));
        Assert.Equal(ClipKind.File, draft.Kind);
        Assert.Equal(["My File.txt", "other.pdf"], draft.FilePaths.Select(ClipClassifier.FileName));
        Assert.Equal("My File.txt\nother.pdf", draft.Title);
        Assert.Equal(2, draft.LineCount);
        Assert.Equal(@"C:\Users\me\Documents", ClipClassifier.ParentFolder(draft.FilePaths[0]));
    }

    [Fact]
    public void MergesMultipleItemsWithoutDuplicatingFormats()
    {
        var snap = Snapshot([[Text("one")], [Text("two"), (ClipboardFormats.Html, HtmlFormat.Encode("<b>two</b>"))]]);
        var draft = Draft(ClipClassifier.Classify(snap));
        Assert.Equal([ClipboardFormats.UnicodeText, ClipboardFormats.Html], draft.Contents.Select(c => c.Format));
        Assert.Equal("one", draft.Title);
    }

    [Fact]
    public void WhitespaceOnlyIsRejected()
    {
        var outcome = ClipClassifier.Classify(Snapshot([[Text("   \n\t")]]));
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(outcome);
        Assert.IsType<CaptureRejection.NothingToStore>(rejected.Rejection);
    }

    [Fact]
    public void MissingDataIsRejected()
    {
        var outcome = ClipClassifier.Classify(Snapshot([[(ClipboardFormats.UnicodeText, null)]]));
        Assert.IsType<CaptureRejection.NothingToStore>(Assert.IsType<CaptureOutcome.Rejected>(outcome).Rejection);
    }

    [Fact]
    public void ConcealedLeavesAGhostEvenWhenOnlyDeclared()
    {
        var declared = new HashSet<string>([ClipboardFormats.UnicodeText, ClipboardFormats.ExcludeFromMonitoring], StringComparer.OrdinalIgnoreCase);
        var snap = Snapshot([[Text("hunter2")]], declared, app: "1Password.exe", appName: "1Password");
        var ghost = Assert.IsType<CaptureOutcome.Ghost>(ClipClassifier.Classify(snap));
        Assert.Equal(new GhostReason.Concealed("1Password"), ghost.Reason);
    }

    [Fact]
    public void CanIncludeInHistoryZeroIsConcealed()
    {
        var snap = Snapshot([[Text("otp 123456")]], canIncludeInHistory: false, appName: "Authenticator");
        var ghost = Assert.IsType<CaptureOutcome.Ghost>(ClipClassifier.Classify(snap));
        Assert.Equal(new GhostReason.Concealed("Authenticator"), ghost.Reason);
    }

    [Fact]
    public void IgnoredCustomFormats()
    {
        var snap = Snapshot([[Text("secret"), ("Clipboard Viewer Ignore", Array.Empty<byte>())]]);
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(snap));
        Assert.Equal(new CaptureRejection.IgnoredFormat("Clipboard Viewer Ignore"), rejected.Rejection);
    }

    [Fact]
    public void IgnoredAppsIncludePasswordManagersByDefault()
    {
        var snap = Snapshot([[Text("pw")]], app: "KeePassXC.exe");
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(snap));
        Assert.Equal(new CaptureRejection.IgnoredApp("KeePassXC.exe"), rejected.Rejection);
        // Exe names are compared case-insensitively.
        Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(Snapshot([[Text("pw")]], app: "bitwarden.exe")));
    }

    [Fact]
    public void IgnoreRegex()
    {
        var policy = CapturePolicy.Default with { IgnoreRegexes = ["^order-[0-9]{6}$"] };
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(Snapshot([[Text("order-123456")]]), policy));
        Assert.IsType<CaptureRejection.MatchedIgnoreRegex>(rejected.Rejection);
        Assert.IsType<CaptureOutcome.Captured>(ClipClassifier.Classify(Snapshot([[Text("order-12")]]), policy));
        // Invalid patterns are ignored rather than breaking capture.
        var broken = CapturePolicy.Default with { IgnoreRegexes = ["("] };
        Assert.IsType<CaptureOutcome.Captured>(ClipClassifier.Classify(Snapshot([[Text("order-12")]]), broken));
    }

    [Fact]
    public void FromNoriIsRejectedWithId()
    {
        var id = Guid.NewGuid();
        var snap = Snapshot([[Text("again"), (ClipboardFormats.NoriItem, Encoding.UTF8.GetBytes(id.ToString()))]]);
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(snap));
        Assert.Equal(new CaptureRejection.FromNori(id), rejected.Rejection);
    }

    [Fact]
    public void DropsDynamicAndMetadataFormats()
    {
        var snap = Snapshot([[
            Text("hello"),
            ("Format123", new byte[] { 1 }),
            ("Embed Source", new byte[] { 2 }),
            ("Chromium Web Custom MIME Data Format", new byte[] { 3 }),
            ("Locale", new byte[] { 4 }),
        ]]);
        var draft = Draft(ClipClassifier.Classify(snap));
        Assert.Equal([ClipboardFormats.UnicodeText], draft.Contents.Select(c => c.Format));
    }

    [Fact]
    public void StripsWordLinkFormats()
    {
        var rtf = Encoding.ASCII.GetBytes(@"{\rtf1 Heading 1}");
        var snap = Snapshot([[
            Text("Heading 1"),
            ("ObjectLink", new byte[] { 1 }),
            ("Link Source", new byte[] { 2 }),
            (ClipboardFormats.Rtf, rtf),
        ]]);
        var draft = Draft(ClipClassifier.Classify(snap));
        Assert.DoesNotContain(draft.Contents, c => c.Format is "ObjectLink" or "Link Source");
        Assert.Contains(draft.Contents, c => c.Format == ClipboardFormats.Rtf);
    }

    [Fact]
    public void PolicyCanDisableImagesFilesAndText()
    {
        var policy = CapturePolicy.Default with { CaptureImages = false, CaptureFiles = false };
        var png = PngHeaderNormalizer.SolidPng(2, 2);
        var snap = Snapshot([[(ClipboardFormats.Png, png), (ClipboardFormats.Rtf, new byte[] { 1 }), Text("caption")]]);
        var draft = Draft(ClipClassifier.Classify(snap, policy));
        Assert.Equal(ClipKind.Text, draft.Kind);
        Assert.Equal([ClipboardFormats.Rtf, ClipboardFormats.UnicodeText], draft.Contents.Select(c => c.Format));

        var noText = policy with { CaptureText = false };
        var rejected = Assert.IsType<CaptureOutcome.Rejected>(ClipClassifier.Classify(Snapshot([[Text("caption")]]), noText));
        Assert.IsType<CaptureRejection.NothingToStore>(rejected.Rejection);
    }

    [Fact]
    public void OversizedImageLeavesAGhost()
    {
        var policy = CapturePolicy.Default with { MaxImageBytes = 4 };
        var png = PngHeaderNormalizer.SolidPng(2, 2);
        var outcome = ClipClassifier.Classify(Snapshot([[(ClipboardFormats.Png, png), Text("alt")]]), policy);
        var ghost = Assert.IsType<CaptureOutcome.Ghost>(outcome);
        Assert.Equal(new GhostReason.ImageTooLarge(png.Length), ghost.Reason);
    }

    [Fact]
    public void OversizedOtherRepresentationIsDroppedButClipKept()
    {
        var policy = CapturePolicy.Default with { MaxOtherRepresentationBytes = 4 };
        var snap = Snapshot([[Text("tiny"), ("Some.Custom.Format", new byte[10]), ("Another", new byte[10])]]);
        var draft = Draft(ClipClassifier.Classify(snap, policy));
        Assert.Equal([ClipboardFormats.UnicodeText], draft.Contents.Select(c => c.Format));
    }

    [Fact]
    public void HugeTextIsTruncated()
    {
        var policy = CapturePolicy.Default with { MaxTextBytes = 16 };
        var snap = Snapshot([[Text(new string('a', 100)), (ClipboardFormats.Rtf, new byte[] { 1 })]]);
        var draft = Draft(ClipClassifier.Classify(snap, policy));
        Assert.True(draft.IsTruncated);
        Assert.Equal(16, draft.CharacterCount);
        Assert.Equal([ClipboardFormats.UnicodeText], draft.Contents.Select(c => c.Format));
    }

    [Fact]
    public void SecretsGoToTheVault()
    {
        var outcome = ClipClassifier.Classify(Snapshot([[Text("AKIAIOSFODNN7EXAMPLE")]]));
        var sensitive = Assert.IsType<CaptureOutcome.Sensitive>(outcome);
        Assert.Equal(SecretDetector.Match.AwsAccessKey, sensitive.Draft.Match);
        Assert.EndsWith("MPLE", sensitive.Draft.Mask);
        Assert.DoesNotContain("AKIA", sensitive.Draft.Mask);

        var policy = CapturePolicy.Default with { MaskSensitive = false };
        Assert.Equal(ClipKind.Text, Draft(ClipClassifier.Classify(Snapshot([[Text("AKIAIOSFODNN7EXAMPLE")]]), policy)).Kind);
    }

    [Fact]
    public void HashIgnoresCustomFormatsButNotContent()
    {
        var a = new List<ClipContent> { ClipDraft.Text("x"), new("Some.Custom", [9]) };
        var b = new List<ClipContent> { ClipDraft.Text("x") };
        var c = new List<ClipContent> { ClipDraft.Text("y") };
        Assert.Equal(ContentHash.Of(a), ContentHash.Of(b));
        Assert.NotEqual(ContentHash.Of(a), ContentHash.Of(c));
    }

    [Fact]
    public void HashIsOrderIndependentAndDistinguishesRich()
    {
        var html = new ClipContent(ClipboardFormats.Html, HtmlFormat.Encode("x"));
        var a = new List<ClipContent> { ClipDraft.Text("x"), html };
        var b = new List<ClipContent> { html, ClipDraft.Text("x") };
        var plain = new List<ClipContent> { ClipDraft.Text("x") };
        Assert.Equal(ContentHash.Of(a), ContentHash.Of(b));
        Assert.NotEqual(ContentHash.Of(a), ContentHash.Of(plain));
        Assert.Equal(64, ContentHash.Of(plain).Length);
    }

    [Fact]
    public void LongTextIsTruncatedForTitleAndSearch()
    {
        var draft = Draft(ClipClassifier.Classify(Snapshot([[Text(new string('a', 20_000))]])));
        Assert.Equal(ClipDraft.MaxTitleLength, draft.Title.Length);
        Assert.Equal(ClipDraft.MaxSearchTextLength, draft.SearchText.Length);
        Assert.Equal(20_000, draft.CharacterCount);
    }

    [Fact]
    public void HtmlOnlyCopyUsesTheHtmlText()
    {
        var html = HtmlFormat.Encode("<p>Hello <b>world</b></p><p>Second</p>");
        var draft = Draft(ClipClassifier.Classify(Snapshot([[(ClipboardFormats.Html, html)]])));
        Assert.Equal(ClipKind.Text, draft.Kind);
        Assert.Equal("Hello world\nSecond", draft.Title);
    }
}

internal static class ClipContentExtensions
{
    public static (string, byte[]?) ToTuple(this ClipContent content) => (content.Format, content.Data);
}
